query_resolver.go

```diff
@@ -0,0 +1,103 @@
+package surrealdb
+
+import (
+	"context"
+	"time"
+
+	"github.com/buger/jsonparser"
+	"github.com/goccy/go-json"
+)
+
+type QueryResult[T any] struct {
+	Result []T    `json:"result"`
+	Status string `json:"status"`
+	Time   string `json:"time"`
+}
+
+type QueryParams = map[string]any
+
+type QueryResolver[T any] struct {
+	err          error
+	response     *RPCRawResponse
+	queryResults []QueryResult[T]
+	query        string
+	params       QueryParams
+	ctx          context.Context
+}
+
+// Query creates a new query resolver
+// It would be ideal if we create a global for *DB
+func Query[T any](db *DB, ctx context.Context, query string, params ...any) *QueryResolver[T] {
+	var paramsData = QueryParams{}
+	if len(params) > 0 {
+		paramsData = params[0].(QueryParams)
+	}
+
+	resolver := &QueryResolver[T]{
+		query:  query,
+		params: paramsData,
+		ctx:    ctx,
+	}
+
+	return resolver.resolve(db)
+}
+
+func (r *QueryResolver[T]) resolve(db *DB) *QueryResolver[T] {
+	ctx, cancel := context.WithTimeout(r.ctx, 5*time.Second)
```

Comment:
This should not be done at the driver level, because if you want to query a lot of data on a slow connection, the driver should not get in your way
That's why we pass `ctx` as the first argument in most methods

Reply:
This is true.... maybe there should be a way to change default timeout(or disable it in place for your own?)

There's cases where RPC returns an error without an ID and go hangs because of the channels, this solves that

I feel it's super cumbersome to create new contexts when making a query 🤔

Reply:
That's why you make a single context per http request/else and just use it everywhere
Not having context means clunky timeouts
If someone want to just not use context, they can make a basic wrapper to abstract it away